rpc/core/tx.go

```diff
@@ -127,9 +139,55 @@ func TxSearch(
 			Index:    r.Index,
 			TxResult: r.Result,
 			Tx:       r.Tx,
-			Proof:    proof,
+			Proof:    txProof,
 		})
 	}
 
 	return &ctypes.ResultTxSearch{Txs: apiResults, TotalCount: totalCount}, nil
 }
+
+func proveTx(height int64, index uint32) (types.TxProof, error) {
+	var (
+		pTxProof tmproto.TxProof
+		txProof  types.TxProof
+	)
+	rawBlock, err := loadRawBlock(env.BlockStore, height)
+	if err != nil {
+		return txProof, err
+	}
+	res, err := env.ProxyAppQuery.QuerySync(abcitypes.RequestQuery{
+		Data: rawBlock,
+		Path: fmt.Sprintf(consts.TxInclusionProofQueryPath, index),
+	})
+	if err != nil {
+		return txProof, err
+	}
+	err = pTxProof.Unmarshal(res.Value)
+	if err != nil {
+		return txProof, err
+	}
+	txProof, err = types.TxProofFromProto(pTxProof)
+	if err != nil {
+		return txProof, err
+	}
+	return txProof, nil
+}
+
+func loadRawBlock(bs state.BlockStore, height int64) ([]byte, error) {
+	var blockMeta = bs.LoadBlockMeta(height)
+	if blockMeta == nil {
+		return nil, fmt.Errorf("no block found for height %d", height)
+	}
+
+	buf := []byte{}
+	for i := 0; i < int(blockMeta.BlockID.PartSetHeader.Total); i++ {
+		part := bs.LoadBlockPart(height, i)
+		// If the part is missing (e.g. since it has been deleted after we
+		// loaded the block meta) we consider the whole block to be missing.
+		if part == nil {
+			return nil, fmt.Errorf("missing block part at height %d part %d", height, i)
+		}
+		buf = append(buf, part.Bytes...)
+	}
+	return buf, nil
```

Comment:
Consider doing this instead:
```go
	block := bs.LoadBlock(height)
	partSet := block.MakePartSet(types.BlockPartSizeBytes)

	buf := []byte{}
	for i := 0; uint32(i) < partSet.Count(); i++ {
		buf = append(buf, partSet.GetPart(i).Bytes...)

	}
	return buf, nil
```

It is less code, less error conditions, and you only access the block store once at the beginning.
I have tested it with the current unit tests and there was no error.